torch/csrc/jit/fuser/compiler.cpp

```diff
@@ -144,6 +148,85 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+// GradSumToSize is only applied to gradient nodes created by autodiff. This is
+// important because it ensures that in the mul and div nodes only one argument
+// (in the case of diff the numerator) has a summed value. If two arguments to
+// mul had one, we would be in trouble, but thanks to the chain rule, we're OK.
+void processGradSumToSize(KernelSpec& spec) {
+  auto graph = spec.graph();
+
+  // keep track of one sumToSize hitting each output
+  auto& outputGradSumToSizes = spec.outputGradSumToSizes();
+  JIT_ASSERT(outputGradSumToSizes.empty());
+  outputGradSumToSizes.resize(graph->outputs().size(), -1);
+
+  static OperatorSet commutes_with_SumToSize{{
+      "aten::mul(Tensor self, Tensor other) -> Tensor",
+      "aten::div(Tensor self, Tensor other) -> Tensor",
+      // for div we might check whether we're the first argument
+      "aten::mul(Tensor self, Scalar other) -> Tensor",
+      "aten::div(Tensor self, Scalar other) -> Tensor",
+      "aten::neg(Tensor self) -> Tensor",
+      "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
+      // add this used to be prim::AutogradAdd
+  }};
+  // Scan the graph. As we will delete nodes, we use the backward ordering.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend();) {
+    auto* node = *it;
+    ++it; // We delete the GradSumToSize nodes, so increment the iterator now.
+    if (node->kind() == aten::_grad_sum_to_size) {
+      use_list uses_to_process(node->output()->uses());
+      while (!uses_to_process.empty()) {
+        auto user = uses_to_process[0].user;
+        auto offset = uses_to_process[0].offset;
+        uses_to_process.erase(uses_to_process.begin());
+        if (user->matches(
+                "aten::type_as(Tensor self, Tensor other) -> Tensor")) {
+          // sometimes, a mask or similar is cast to the same
+          // type as the gradient. But we only want to allow that as offset==1
+          // (i.e. the type)
+          JIT_ASSERT(offset == 1);
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (commutes_with_SumToSize.find(user)) {
+          // these are expressions that might occur during autotdiff operating
+          // on the gradient (matmul would likely be, too but we don't fuse it)
+          // note that for mul, we know (from the chain rule) that only one
+          // factor will be stemming from a calculation involving gradients so
+          // we know that we can move SumToSize across it
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (user->kind() == aten::_grad_sum_to_size) {
+          // we can just keep the last, so we're done here (this case might not
+          // happen due to how we organize the search and remove nodes...)
```

Comment:
Well, because we're iterating in reverse topological order, it sounds like this shouldn't happen. It might be better to `JIT_ASSERT(false)` here?

Reply:
OK, I'll try that.